### PR TITLE
Add reCAPTCHA v3 to Braze Preference Center form submission.

### PR DESCRIPTION
**Pull request by Shinsina**
Mostly copied logic from https://github.com/parameter1/base-cms/blob/master/packages/marko-web-contact-us/browser/form.vue was able to confirm it will fail (forgot to include our domain (parameter1.com which is used for the development instance) on the acceptable domain list initially)

WORKS

### CODE REVIEW

**Comment by Shinsina** [comment]
[Add reCAPTCHA v3 to Braze preference center feature](https://trello.com/c/cgGnRbda/82-add-recaptcha-v3-to-braze-preference-center-feature)

File: packages/global/browser/braze-preference-center.vue [hunk 1]
```diff
@@ -141,21 +161,29 @@ export default {
...
-          method: 'post',
-          headers: { 'content-type': 'application/json' },
-          body: JSON.stringify({ email: this.email, optIns: this.optIns }),
-        });
-        if (!r.ok) {
-          const { error } = await r.json();
-          if (error) throw new Error(error);
-          throw new Error(`${r.status} ${r.statusText}`);
+      this.error = null;
+      const token = await recaptchaGetToken({ siteKey: this.siteKey, action: 'brazerPreferenceCenter' });
```
Comment (solocommand) [inline packages/global/browser/braze-preference-center.vue:165]: This should be `brazePreferenceCenter` not `brazerPreferenceCenter`

File: packages/global/browser/braze-preference-center.vue [hunk 2]
```diff
@@ -141,21 +161,29 @@ export default {
...
+
+      if (token) {
```
Comment (solocommand) [inline packages/global/browser/braze-preference-center.vue:167]: Don't wrap all of this in an if check. Instead, put your recaptcha call inside the try, and throw if there is no token. The reset of the error message is fine to keep.

File: packages/global/browser/braze-preference-center.vue [hunk 3]
```diff
@@ -167,6 +195,17 @@ export default {
...
+      try {
+        this.recaptcha.loading = true;
+        this.recaptcha.error = null;
+        await recaptchaLoad({ siteKey: this.siteKey });
+      } catch (e) {
+        this.recaptcha.error = e;
+      } finally {
+        this.recaptcha.loading = false;
+      }
+    },
```
Comment (solocommand) [inline packages/global/browser/braze-preference-center.vue:208]: No need for separate and error/loading states here. 
```suggestion
    async loadRecaptcha() {
      try {
        this.loading = true;
        await recaptchaLoad({ siteKey: this.siteKey });
      } catch (e) {
        this.error = e.message;
      } finally {
        this.loading = false;
      }
    },
```

File: packages/global/routes/braze.js [hunk 4]
```diff
@@ -62,8 +64,9 @@ module.exports = (app) => {
   app.post('/user/subscribe', json(), asyncRoute(async (req, res) => {
     try {
       const { body } = req;
-      const { email, optIns } = body;
+      const { email, optIns, token } = body;
 
+      await validateToken({ token, secretKey: RECAPTCHA_V3_SECRET_KEY, actions: ['brazerPreferenceCenter'] });
```
Comment (solocommand) [inline packages/global/routes/braze.js:69]: Rename action key

File: packages/global/browser/braze-preference-center.vue [hunk 5]
```diff
@@ -104,13 +120,17 @@ export default {
     didSubmit: false,
     optIns: {},
     timeout: null,
+    recaptcha: { loading: false, error: null },
```
Comment (solocommand) [inline packages/global/browser/braze-preference-center.vue:123]: remove

File: packages/global/browser/braze-preference-center.vue [hunk 6]
```diff
@@ -50,6 +54,12 @@
       <p v-if="error" class="mt-3 text-danger">
         An error occurred while processing your request: {{ error }}.
       </p>
+      <p
+        v-if="recaptcha.error"
+        class="mt-3 text-danger"
+      >
+        A recaptcha error occurred {{ recaptcha.error.message }}
+      </p>
```
Comment (solocommand) [inline packages/global/browser/braze-preference-center.vue:62]: Remove

File: packages/global/browser/braze-preference-center.vue [hunk 7]
```diff
@@ -41,7 +41,11 @@
           />
         </div>
       </fieldset>
-      <button type="submit" class="btn btn-primary" :disabled="loading">
+      <button
+        type="submit"
+        class="btn btn-primary"
+        :disabled="loading || recaptcha.loading || recaptcha.error"
+      >
```
Comment (solocommand) [inline packages/global/browser/braze-preference-center.vue:48]: Revert

**Review by solocommand — CHANGES_REQUESTED** [review]

**Review by solocommand — APPROVED** [review]

**Pull request merged**